### PR TITLE
BUG: Fix face orientations of bar3d

### LINKED ISSUE #12138: Faces of Axes3d.bar3d are not oriented correctly
All of the faces of the polygons created by `bar3d` are oriented (using a right-hand rule on the vertex order) in the positive axis directions. As a result, the shading used by `shade=True` is incorrect, as it shades both the front and back faces with the same brightness.

Rotating the [`hist3d`](https://matplotlib.org/examples/mplot3d/hist3d_demo.html) demo, we can see the left and right faces of the bars are the same color:

image

image

The fix is fairly straightforward, thankfully

### PR DESCRIPTION
**Pull request by eric-wieser**
Fixes #12138, which is caused by these incorrect orientations

I was hoping I could modify the existing bar3d test with something like:
```python
@pytest.mark.parametrize('azim', [-60, 60])
@image_comparison(
    baseline_images=['bar3d_shaded'],
    remove_text=True,
    extensions=['png']
)
def test_bar3d_shaded(azim):
    ...
    ax.view_init(azim=azim)
```
but I couldn't work out how to make `parametrize` work with the `baseline_images` argument of `image_comparison`

### CODE REVIEW

**Comment by eric-wieser** [comment]
Worked out how to parametrize the test.

Let me know if you'd like to see the test in a separate PR, to make the result easier to compare.

**Comment by dstansby** [comment]
Would it be possible to stick the test images on different axes on the same figure (something like `subplots(2, 2)`? This will save a few images, and also make a nice side by side comparison of the different viewing angles.

**Comment by eric-wieser** [comment]
I avoided that because it leads to a 100% diff in the target image, and also removes the ability to use pytest.parametrize.

However, I'm not super committed to that decision - if you think that'd be best, can you comment to that effect on #12273?

**Comment by eric-wieser** [comment]
Updated. Something odd seems to be happening though - the light source is above the plot, yet it renders as if it were below.

Might also be worth considering changing the default lighting angle so that it doesn't hit cuboids at exactly 45 degrees causing adjacent faces to be the same color.

**Comment by ImportanceOfBeingErnest** [comment]
I would have thought that the reason for the two bright faces is that, independent of the viewing angle, you have one bright face facing the viewer. Is this undesireable?  
I think it would be good to see the same case visually with a before/after - comparisson here.

**Comment by eric-wieser** [comment]
@ImportanceOfBeingErnest: you have that before/after comparision in the diff of the result image

**Comment by eric-wieser** [comment]
Ugh, I didn't see you'd replied here.

> Somehow undesired in my eyes, I would much rather have ...

I agree, that would be a neat feature. I'm working on adding shading to some extra plots- once that's done, maybe I'll investigate a fixed light source.

> Would there be any way to make it consistent for the zero-extent bar as well?

That's a bug I'd rather not deal with yet that comes down to z-sorting. We only got away with it before because the upper and lower faces were the same color. I'd prefer to leave that to a later PR.

File: lib/mpl_toolkits/mplot3d/axes3d.py [hunk 1]
```diff
@@ -1717,13 +1717,16 @@ def _generate_normals(self, polygons):
...
         '''
 
         normals = []
         for verts in polygons:
-            v1 = np.array(verts[0]) - np.array(verts[1])
-            v2 = np.array(verts[2]) - np.array(verts[0])
-            normals.append(np.cross(v1, v2))
+            v1 = np.array(verts[1]) - np.array(verts[0])
+            v2 = np.array(verts[0]) - np.array(verts[2])
+            normals.append(np.cross(v2, v1))
```
Comment (anntzer) [inline lib/mpl_toolkits/mplot3d/axes3d.py]: Doesn't really matter, but given that you mention the right-hand rule it seems most natural for me to put verts[0] at my palm, verts[1] at the extremity of my thumb and verts[2] of my index, and then the normal is `np.cross(v1-v0, v2-v0)` (which is equal to what you wrote, but perhaps easier to follow).

**Comment by anntzer** [comment]
Why did the notshaded baseline image change?

File: lib/mpl_toolkits/mplot3d/axes3d.py (hunk 1 above)
Comment (eric-wieser) [reply lib/mpl_toolkits/mplot3d/axes3d.py]: I was trying to formulate this as a right hand screw rule, but what you have there makes sense too.

Do you want me to change that here, or shall I just do so when I remove this code in #12136

**Comment by eric-wieser** [comment]
> Why did the notshaded baseline image change?

Because some faces are now rendered with the points in reverse order, and I guess the renderer somehow cares about that.

File: lib/mpl_toolkits/mplot3d/axes3d.py (hunk 1 above)
Comment (anntzer) [reply lib/mpl_toolkits/mplot3d/axes3d.py]: Can you change it here, and also the corresponding (vectorized) code in #12136?

Comment (eric-wieser) [reply lib/mpl_toolkits/mplot3d/axes3d.py]: Done

**Comment by eric-wieser** [comment]
Not sure why CI fails

**Comment by eric-wieser** [comment]
Looks like it maybe just needs rebasing on top of a commit with CI configured. Shall I go ahead and do that?

**Comment by eric-wieser** [comment]
@dstansby: Rebased

**Comment by eric-wieser** [comment]
Was hoping you'd merge rather than squash to keep me from having to rebase. Oh well - thanks for putting this in anyway!

**Comment by QuLogic** [comment]
That setting is sticky, so I don't know if @timhoffm did that on purpose (I wouldn't have squashed for this PR). If it wasn't on purpose, please @timhoffm be careful on the next merge.

**Comment by eric-wieser** [comment]
The stickiness has caught me out in the past too.

**Comment by QuLogic** [comment]
Hmm, that probably should be announced. I'm also not particularly for it when it's an experienced contributor.

**Comment by timhoffm** [comment]
Sorry if squashing caused troubles. The changes seemed rather small and related so that I did not consider it important to keep separate commits.

+1 that an author should explicitly mention that commits should not be squashed.

**Pull request merged**